### DISCUSSION TITLE
Are Power and Yield values match the inverter output or we must account for some loss ?

### DISCUSSION THREAD

**Discussion by mathieucarbou** [tbnobody/OpenDTU#476] [Q&A]
Hello,

I have installed OpenDTU, this is awesome. But I see that the power / yield values do not match what my Shelly device reports for energy and power. The Shelly device report a little less. But I know that the Shelly devices have also some margins of error.

If I want to get the most accurate values in Home  Assistant, I am wondering if I should take the values from the Shelly device or from the DTU. I don't know how accurate the values from the DTU are ?

Example: regarding the power value from the DTU: is it a theoretical power value converted from the AC input, not accounting for any loss, or it is really the real power injected in the wires from the inverters ?

Thanks!

**Comment by ahaspharos** [comment]
What shelly device are you using? the only shelly product for "measuring" ist the EM series. i have an EM3 installed wich i use to set the power limit with the openDTU and an EM directly at the HM-800 for measruring solar power.
The results are really close.
Shelly PMs are no measuring devices, they are "guessing"

**Comment by mathieucarbou** [comment]
Thanks for your reply!
My solar panels are plugged in an AC socket (micro-inverters). But there are other consumers on this circuit so when I tried the EM, it was reporting the difference of production/consumption. So that is why I have changed for a Shelly Plug 2 for the solar panel AC plug.

But based on your answer, I should understand that the power reported by OpenDTU is the real power sent on the AC wire ?

**Reply by ahaspharos** [reply]
that's what my shelly em is telling me. i have a difference <2%

**Comment by mathieucarbou** [comment]
Difference is 5-7% for me... That's a lot.... I have the possibility to use either Shelly Plug, Plug S or Plus 1PM. I don't know which one of them is most accurate though. I will ask or test. Peak production is 1600W.

**Reply by Solarteur** [reply]
Maybe the post here
https://github.com/tbnobody/OpenDTU/discussions/394#discussioncomment-4251312

 is helpful, including the section on the metering accuracy of "official" electricity meters.

**Reply by mathieucarbou** [reply]
> Maybe the post here [#394 (comment)](https://github.com/tbnobody/OpenDTU/discussions/394#discussioncomment-4251312)
> 
> is helpful, including the section on the metering accuracy of "official" electricity meters.

Thanks!

FYI, when discussing that with another person, I also found that the power reported by the shelly plug is roughly equal to the efficiency of the inverter by the power reported by the inverter.

But since there are error margins pretty much everywhere (inverter stats, shelly stats) then sometimes it worked, sometimes not. So I was wondering what's the right way to interpret these stats.

Translating the ticket linked, I am not sure either than I can have a definitive answer: it seems more something someones else also determined ?

**Comment by mathieucarbou** [comment]
For ref: https://www.facebook.com/groups/ShellyIoTCommunitySupport/permalink/5547181905381128/

**Comment by mathieucarbou** [comment]
To close the conversation, here is how I solved it.

Turns out the actual equation is:

`power seen by shelly plug` * `shelly plug error rate` = `power shown on dtu` * `efficiency shown on dtu`


This link - provided by @Solarteur - helped me (https://github.com/tbnobody/OpenDTU/discussions/394#discussioncomment-4251312)

In Home Assistant, I have creating sensors for each inverter, computing the real power:

```
      - name: hoymiles_hm_400_1_power_real
        unique_id: "01e37eed-3045-4eee-a786-1249567fe311"
        state_class: measurement
        device_class: power
        unit_of_measurement: W
        state: "{{ (states('sensor.hoymiles_hm_400_1_power')|float * states('sensor.hoymiles_hm_400_1_efficiency')|float / 100.0) | round(2) }}"
      - name: hoymiles_hm_400_2_power_real
        unique_id: "01e37eed-3045-4eee-a786-1249567fe312"
        state_class: measurement
        device_class: power
        unit_of_measurement: W
        state: "{{ (states('sensor.hoymiles_hm_400_2_power')|float * states('sensor.hoymiles_hm_400_2_efficiency')|float / 100.0) | round(2) }}"
      - name: hoymiles_hm_400_3_power_real
        unique_id: "01e37eed-3045-4eee-a786-1249567fe313"
        state_class: measurement
        device_class: power
        unit_of_measurement: W
        state: "{{ (states('sensor.hoymiles_hm_400_3_power')|float * states('sensor.hoymiles_hm_400_3_efficiency')|float / 100.0) | round(2) }}"
      - name: hoymiles_hm_400_4_power_real
        unique_id: "01e37eed-3045-4eee-a786-1249567fe314"
        state_class: measurement
        device_class: power
        unit_of_measurement: W
        state: "{{ (states('sensor.hoymiles_hm_400_4_power')|float * states('sensor.hoymiles_hm_400_4_efficiency')|float / 100.0) | round(2) }}"
```

Than, I have created a sensor accounting for my plug error rate:

```
      - name: sunology_play_power
        unique_id: "01e37eed-3045-4eee-a786-1249567fe300"
        state_class: measurement
        device_class: power
        unit_of_measurement: W
        state: "{{ (states('sensor.sunology_play_plug_power')|float * 0.975) | round(1)}}" # 2.5% error-rate of shelly plug
```

I have created also an entry which is the sum of all 4 `*_power_real` sensor above.

Thanks to that, I have the exact same power number both reported from shelly plug and DTU:

![Screenshot 2023-01-03 at 13 17 58](https://user-images.githubusercontent.com/61346/210355978-9abe3f96-3cdc-43f9-a9fd-48610b1d7eb9.png)

![Screenshot 2023-01-03 at 13 18 06](https://user-images.githubusercontent.com/61346/210355982-6ddd2c53-e0c2-4b35-be8e-e17969c96558.png)

The same rule holds true for the energy accumulated (YieldDay for DTU and energy grabbed from shelly plug)

**Reply by ahaspharos** [reply]
as i understand the dtu efficiency ist the conversion factor from dc to ac 
![image](https://user-images.githubusercontent.com/8578356/210368317-8cf33eb2-f294-4cdc-aa4d-29097b172006.png)
so why are you multiplying power an efficiency? it's already the product of power dc and efficiency, and accurate.

**Reply by mathieucarbou** [reply]
Hum... That's right... So the error rate is a lot for my EM and plugs...

**Reply by Solarteur** [reply]
This applies to the instantaneous power (W). The feed-in (Wh/kWh, day or total) is retrieved from the inverter as DC values.

**Reply by mathieucarbou** [reply]
Ok I understand a bit more now ;-) So in summary:
- Power is already computed with the efficiency
- Daily energy (YieldDay) comes from DC so we have to apply it an efficiency factor (roughly 95.2% in my case)
